lte/gateway/c/core/oai/tasks/grpc_service/S1apServiceImpl.cpp

```diff
@@ -16,54 +16,51 @@
  */
 #include "lte/gateway/c/core/oai/tasks/grpc_service/S1apServiceImpl.h"
 
+#include <memory>
 #include <string>
 
-extern "C" {
-#include "lte/gateway/c/core/oai/common/assertions.h"
-#include "lte/gateway/c/core/oai/lib/hashtable/hashtable.h"
-#include "lte/gateway/c/core/oai/common/log.h"
-#include "lte/gateway/c/core/oai/include/s1ap_state.h"
-}
+#include "lte/protos/oai/s1ap_state.pb.h"
+#include "orc8r/protos/common.pb.h"
+#include "redis_utils/redis_client.h"
 
 using grpc::ServerContext;
 using grpc::Status;
-using magma::EnbStateResult;
-using magma::S1apService;
+using magma::lte::EnbStateResult;
+using magma::lte::S1apService;
+using magma::lte::oai::S1apState;
+using magma::orc8r::Void;
 
 namespace magma {
-using namespace lte;
-using namespace orc8r;
+namespace lte {
+
+S1apServiceImpl::S1apServiceImpl() : client_(nullptr) {}
 
-S1apServiceImpl::S1apServiceImpl() {}
+void S1apServiceImpl::init(std::shared_ptr<RedisClient> client) {
+  client_ = client;
+}
 
 Status S1apServiceImpl::GetENBState(
     ServerContext* context, const Void* request, EnbStateResult* response) {
   OAILOG_DEBUG(LOG_UTIL, "Received GetENBState GRPC request\n");
 
-  // Get state from S1APStateManager
-  // TODO: Get state through ITTI message from S1AP task, as it's read only
-  // it will not affect ownership
-  s1ap_state_t* s1ap_state = get_s1ap_state(false);
-  if (s1ap_state != nullptr) {
-    hashtable_rc_t ht_rc;
-    hashtable_key_array_t* ht_keys = hashtable_ts_get_keys(&s1ap_state->enbs);
-    if (ht_keys == nullptr) {
-      return Status::OK;
-    }
+  S1apState s1ap_state_msg = S1apState();
+  auto rc                  = client_->read_proto("s1ap_state", s1ap_state_msg);
+  if (rc != RETURNok) {
+    OAILOG_DEBUG(
+        LOG_UTIL,
+        "Failure while reading s1ap_state from redis. Cancelling GetENBState "
+        "request");
+    return Status::CANCELLED;
+  }
```

Comment:
This type of design has two major issues:

1. GRPC servicer is not the owner of the state and it should not rely on Redis backend to fetch it. The actual owner may evolve and change how/when these states are updated/refactored on Redis.

2. This breaks the parity between stateful and stateless. E.g., if someone disables stateless feature, eNB state will no longer be read.